pkg/errors/errors.go

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+//nolint
+// The idea of writing errors package in thanos is highly motivated from the Tast project of Chromium OS Authors. However, instead of
+// copying the package, we end up writing our own simplified logic borrowing some ideas from the errors and github.com/pkg/errors.
+// A big thanks to all of them.
+
+// Package errors provides basic utilities to manipulate errors with a useful stacktrace. It combines the
+// benefits of errors.New and fmt.Errorf world into a single package.
+package errors
+
+import (
+	//lint:ignore faillint Custom errors package needs to import standard library errors.
+	"errors"
+	"fmt"
+	"strings"
+)
+
+// base is the fundamental struct that implements the error interface and the acts as the backbone of this errors package.
+type base struct {
+	// info contains the error message passed through calls like errors.Wrap, errors.New.
+	info string
+	// stacktrace stores information about the program counters - i.e. where this error was generated.
+	stack stacktrace
+	// err is the actual error which is being wrapped with a stacktrace and message information.
+	err error
+}
+
+// Error implements the error interface.
+func (b *base) Error() string {
+	if b.err != nil {
+		return fmt.Sprintf("%s: %s", b.info, b.err.Error())
+	}
+	return b.info
+}
+
+// Unwrap implements the error Unwrap interface.
+func (b *base) Unwrap() error {
+	return b.err
+}
+
+// Format implements the fmt.Formatter interface to support the formatting of an error chain with "%+v" verb.
+// Whenever error is printed with %+v format verb, stacktrace info gets dumped to the output.
+func (b *base) Format(s fmt.State, verb rune) {
+	if verb == 'v' && s.Flag('+') {
+		s.Write([]byte(formatErrorChain(b)))
+		return
+	}
+
+	s.Write([]byte(b.Error()))
+}
+
+// New returns an error that formats as the given text and a stacktrace with recent
+// call frames. Each call to New returns a distinct error value even if the text is
+// identical. An alternative to errors.New function.
+func New(msg string) error {
+	return &base{
+		info:  msg,
+		stack: newStackTrace(),
+		err:   nil,
+	}
+}
+
+// Errorf formats according to a format specifier and returns a new error with
+// a stacktrace containing recent call frames. An alternative to fmt.Errorf function.
+func Errorf(format string, args ...interface{}) error {
+	return &base{
+		info:  fmt.Sprintf(format, args...),
+		stack: newStackTrace(),
+		err:   nil,
+	}
+}
+
+// Wrap returns a new error with the supplied message as an argument and wrapping another error
+// with a stacktrace containing recent call frames.
+//
+// If cause is nil, this is the same as New.
+func Wrap(cause error, msg string) error {
```

Comment:
I see we left our both non `f` and `f` methods at the end? 🤔

Reply:
Hey, I thought we are focussing on end-users experience that's why ended up keeping both.
I have changed my mind, let's simplify things and stick to only the `f` versions :smile:

Reply:
Mission accomplished : )
![image](https://user-images.githubusercontent.com/41498427/163520038-abc9e285-6d50-463c-b191-1e88da27b883.png)